### ISSUE TITLE
Bug: [no-restricted-imports] allowTypeImports allows side effect imports

### ISSUE THREAD

**Issue by joshkel** [typescript-eslint/typescript-eslint#7559]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.2.2&fileType=.tsx&code=JYWwDg9gTgLgBAIgGYQgg3AKFJWcYCeYApnAN5wBiqcAvnElBCIimljtPBYSXAEIBDKHQZMWyVBmzgu5AYIBeoxs1ZT0QA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1uhXyrPyIAJrUoBbYh2j5UGANqRE0aFMgAaTOGw5iAQ3wALWWDlbtELOe2QmusYnRQAZhw7qzVqLvjwOAdwAVEkQASQkpGUc%2BWEQPbQBfOLAAXQ941IhE%2BKA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```typescript
import "foo";
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-restricted-imports": ["error", {
      "paths": [
        {
          "name": "foo",
          "allowTypeImports": true
        }
      ]
    }]
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

The purpose of [type-only imports](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-8.html#type-only-imports-and-export) is to ensure that an import has no runtime behavior, so I would expect that the [allowTypeImports](https://typescript-eslint.io/rules/no-restricted-imports#allowtypeimports) option disallows [side effect imports](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import#import_a_module_for_its_side_effects_only) such as `import 'foo'`.

### Actual Result

Side effect imports are allowed.

### Additional Info

This issue was apparently introduced in typescript-eslint/eslint-plugin 6.3.0, as a result of https://github.com/typescript-eslint/typescript-eslint/pull/7379.
